### ISSUE TITLE
Fix All in Document, Project, Solution (Test?)

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/roslyn#22591]
An interesting side discussion from PR #19999, around a test project to test fix all, and suitable candidate for the codefix to use. 

> @paul1956 **typed**
>An interesting test would be to run this codefix on a copy of Roslyn source "Solution" because I believe you will get an OutOfMemory exception. Yet this is a perfect example of where a solution wide CodeFix is needed.
>
> @AdamSpeight2008 **typed**
> @paul1956 It would be a good test. You'll be able to see the smoke from VS for miles.
>
>@AdamSpeight2008 isn't that a problem? If there is a limit of X fixes before bad things happen how about limiting code fixes to a smaller number or 75% of available memory, or fix the algorithm to aggregate each document. I would be much happier if a code fix stopped after 50 (100, 500...) fixes (and I could run it again) which might be one fix in each of 50 files or 50 fixes in one document, then what I have today, where I have to search for the error, go to the file, fix all the issues then find another file (which might have 1 issue), and repeat for each file. Switching from Implicit to Explicit declarations in just the Roslyn VB CodeAnalysis project is 39,000 changes and going the other way is even more. Changing to Optional Default Parameter is around 6000 changes in ~800 files (very rough guess).
>

**Related To:**  #22539

**Comment by AdamSpeight2008** [comment]
The original analysis used to gauge the potential usefulness of the proposed feature and code fix, showed 2229 out of 2612 (>85%) from a scan of the Rosyln repo.

**Comment by AdamSpeight2008** [comment]
A possible solution would be to reuse the same document for the next fix, akin to Enumerable.Aggregate.
```vbnet
UpdatedDocument = CurrentDocument
UpdatedDocument = CodeFixes.Aggregate( UpdatedDocument,
                                       Function( PreviousFixedDocument, ThisCodeFix) CodeFix(PreviousFixedDocument) )
CurrentDocument = UpdatedDocument
```
So there is only at most one new copy of updated document during the application of the fixes.

**Comment by sharwell** [comment]
When working on StyleCop Analyzers, our experience there was the batch fixer was not particularly suitable for Fix All operations. It provided an initial approach, but we considered the proper approach to be implementing a dedicated Fix All operation for each analyzer. The difference between the approaches was so pronounced that we ended up using a ⌛️ instead of a ✔️ on the [StyleCop Analyzers status page](https://stylecop.pdelvo.com) to mark code fixes that use the batch fixer.

In other words, for that project we don't make a claim that Fix All is supported if it relies on the batch fixer.

**Comment by paul1956** [comment]
If the batch fix just monitored memory and stopped before running out that would be acceptable since large fixes usually require multiple passes to fix all the issues in many cases anyway. Merging into a single document would also work for fixes that don't have the effect to creating or fixing additional issues. For 1.0 of Roslyn I did implement my own batch fixer but there is a lot of code to get right and do we really want to everyone to have to duplicate that code. Maybe we could have a few different batch fixers that developers could select from based on their needs or extract the source of a standalone batch fixer that developers could include in their code.

**Comment by caseytp** [comment]
Can anyone confirm if this is still an issue?

**Comment by CyrusNajmabadi** [comment]
I rewrote the default batch-fixer to be a lot smarter and more efficient.  I cannot guarantee that it won't run out of memory applying something absolutely massive. I will say that i tested it on Roslyn with several solution-wide fixes and it had no problem.
